principles.rst

```diff
@@ -168,26 +234,38 @@ Lexical Scoping Principle (LSP)
 **Principle**:
 
 a. For every appearance of an identifier,
-it is possible to determine whether that appearance is a *binding site* or an *occurrence* without examining the context.
+   it is possible to determine whether that appearance is a mere *occurrence*,
+   and thus must be bound elsewhere for the program to be valid,
+   or the variable is a *binding site* (or causes an implicit binding, which is close enough),
+   without examining the context.
 
 b. For every *occurrence* of an identifier,
-it is possible to uniquely identify its *binding site*, without involving the type system.
+   it is possible to uniquely identify its *binding site*, without involving the type system.
+
+This builds upon the `Explicit Binding Principle`_:
+whereas that former principle ensures that explicit alternatives to implicit binding constructs *exist at all*,
+this latter principle makes those explicit alternatives *compulsory*, because we must not have implicit binding in order to uphold this principle.
 
 The `Lexical Scoping Principle`_ is almost true today, with the following nuances:
 
-1. Template Haskell splices may need to be run before completing name resolution (and running those splices requires type-checking them).
+#. Template Haskell splices may need to be run before completing name resolution (and running those splices requires type-checking them).
 
-2. The `deprecated mechanism <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/exts/duplicate_record_fields.html#selector-functions>`_ for disambiguating duplicate record fields violates the `Lexical Scoping Principle`_ by requiring the type system.
+#. The `deprecated mechanism <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/exts/duplicate_record_fields.html#selector-functions>`_ for disambiguating duplicate record fields violates the `Lexical Scoping Principle`_ by requiring the type system.
 
-3. In a pattern signature,
+#. In a pattern signature,
    if we have ``f (x :: Maybe a)``,
    the ``a`` is an occurrence if ``a`` is already in scope,
-   and it is a binding site otherwise.
+   and is implicitly bound otherwise.
+
+#. In a type signature,
+   if we have ``f :: a -> a``,
+   the ``a`` is an occurrence if ``a`` is already in scope,
+   and is implicitly bound otherwise.
```

Comment:
See what I wrote below in https://github.com/ghc-proposals/ghc-proposals/pull/532#discussion_r1035161855 . The **Explicit Binding Principle** is was bit wish-washy before on whether it or the **Lexical Scoping Principle** said it was no good. But now I am making the Explicit Binding Principle be about something else, "explicit *and unambiguous*" bindings, so it squarely falls under its purview.

-------

A natural follow-up question might be "that's great but who cares which principle catches the violation? Why does it matter?"

Well, my next thought it that banning implicit bindings, like promoting pun-free code (#270) is probably controversial. However, what would be less controversial is saying we have explicit option to teach/explain/desugar all our implicit options.

The **Explicit Binding Principle** I am trying to make the first, not second, principle, because it alone should be about that point. It (along with the Explicit Variable Principle, which is almost the same), says we need to have `\ @a -> ..`, `\(DataCon @a) -> ...`, `let type a = _ in ...`, etc. so long as we have their implicit counterparts. This I hope is not controversial! We should have syntax to be able state exactly what we mean. I would hope we could even standardize many of these at some point in a future language report.

The **Lexical Scoping Principle**, by contrast, has the controversial residual, which is merely *having* explicit counterparts isn't good enough, and for maximally easy *reading* we need to also be able to ensure the implicit/ambiguous ones are not used. I personally wouldn't mind if `-XImplicitBinds` went away and the language just became Agda style where we must be explicit, but I don't expect the Haskell community writ large to be persuaded of this any time soon. So the **Lexical Scoping Principle** keeps the controversial bits and tempers them down to being we should merely have the *option* to disable implicit/ambiguous constructs.